src/node_blob.cc

```diff
@@ -150,70 +245,128 @@ void Blob::ToSlice(const FunctionCallbackInfo<Value>& args) {
+           std::shared_ptr<DataQueue> data_queue)
+    : BaseObject(env, obj), data_queue_(data_queue) {
+  MakeWeak();
+}
 
-  for (const auto& entry : entries()) {
-    if (start + entry.offset > entry.store->ByteLength()) {
-      start -= entry.length;
-      continue;
-    }
+Blob::Reader::Reader(Environment* env,
+                     v8::Local<v8::Object> obj,
+                     BaseObjectPtr<Blob> strong_ptr)
+    : AsyncWrap(env, obj, AsyncWrap::PROVIDER_BLOBREADER),
+      inner_(strong_ptr->data_queue_->getReader()),
+      strong_ptr_(std::move(strong_ptr)) {
+  MakeWeak();
+}
 
-    size_t offset = entry.offset + start;
-    size_t len = std::min(remaining, entry.store->ByteLength() - offset);
-    slices.emplace_back(BlobEntry{entry.store, len, offset});
+bool Blob::Reader::HasInstance(Environment* env, v8::Local<v8::Value> value) {
+  return GetConstructorTemplate(env)->HasInstance(value);
+}
 
-    remaining -= len;
-    start = 0;
+Local<FunctionTemplate> Blob::Reader::GetConstructorTemplate(Environment* env) {
+  Local<FunctionTemplate> tmpl = env->blob_reader_constructor_template();
+  if (tmpl.IsEmpty()) {
+    Isolate* isolate = env->isolate();
+    tmpl = NewFunctionTemplate(isolate, nullptr);
+    tmpl->InstanceTemplate()->SetInternalFieldCount(
+        BaseObject::kInternalFieldCount);
+    tmpl->Inherit(BaseObject::GetConstructorTemplate(env));
+    tmpl->SetClassName(FIXED_ONE_BYTE_STRING(env->isolate(), "BlobReader"));
+    SetProtoMethod(env->isolate(), tmpl, "pull", Pull);
+    env->set_blob_reader_constructor_template(tmpl);
+  }
+  return tmpl;
+}
 
-    if (remaining == 0)
-      break;
+BaseObjectPtr<Blob::Reader> Blob::Reader::Create(Environment* env,
+                                                 BaseObjectPtr<Blob> blob) {
+  Local<Object> obj;
+  if (!GetConstructorTemplate(env)
+           ->InstanceTemplate()
+           ->NewInstance(env->context())
+           .ToLocal(&obj)) {
+    return BaseObjectPtr<Blob::Reader>();
   }
 
-  return Create(env, slices, total);
+  return MakeBaseObject<Blob::Reader>(env, obj, std::move(blob));
 }
 
-Blob::Blob(
-    Environment* env,
-    v8::Local<v8::Object> obj,
-    const std::vector<BlobEntry>& store,
-    size_t length)
-    : BaseObject(env, obj),
-      store_(store),
-      length_(length) {
-  MakeWeak();
+void Blob::Reader::Pull(const FunctionCallbackInfo<Value>& args) {
+  Environment* env = Environment::GetCurrent(args);
+  Blob::Reader* reader;
+  ASSIGN_OR_RETURN_UNWRAP(&reader, args.Holder());
+
+  CHECK(args[0]->IsFunction());
+  Local<Function> fn = args[0].As<Function>();
+  CHECK(!fn->IsConstructor());
+
+  if (reader->eos_) {
+    Local<Value> arg = Int32::New(env->isolate(), bob::STATUS_EOS);
+    reader->MakeCallback(fn, 1, &arg);
+    return args.GetReturnValue().Set(bob::STATUS_EOS);
+  }
+
+  struct Impl {
+    BaseObjectPtr<Blob::Reader> reader;
+    Global<Function> callback;
+    Environment* env;
+  };
+  Impl* impl = new Impl();
```

Comment:
Why not make this a `std::unique_ptr` already? That’s more obviuosly correct and doesn’t seem to have any downsides

Reply:
Added a comment on this. I'll revisit this separate but making this a `std::unique_ptr` causes the later std::move of the lambda complain that it is not copy constructible.